### PR TITLE
2.0.0

### PR DESCRIPTION
**Pull request by hgromer**

### CODE REVIEW

File: ecabc/bees.py [hunk 1]
```diff
@@ -2,46 +2,67 @@
...
-    def getFitnessScore(self, values, fitnessFunction):
-        if self.beeType != "employer":
-            raise RuntimeError("Cannot get fitness score on a non-employer bee")
-        else:
-            # Your fitness function must take a certain set of values that you would like to optimize
-            fitnessScore = fitnessFunction(values)  
-            if self.currFitnessScore == None or fitnessScore < self.currFitnessScore:
-                self.value = values
-                self.currFitnessScore = fitnessScore
+    def __init__(self, values=[]):
```
Comment (hgromer) [inline ecabc/bees.py:21]: Avoided using the .__ prefix, unsure if I should add getters and mark them as private instead.

**Comment by tjkessler** [comment]
The Settings module/class seems unnecessary - value ranges and number of employers can be handed to the ABC when it's initialized, minimize can be a @property with setter methods, and best score/values can be @properties. I currently do not see a way to access best score/values directly from the ABC, making higher-level analysis of results difficult, as you would need to rely on opening the settings file to access results.

** Update, I see the get_best_performer method handles getting the best score/values. Having them as properties so they wouldn't need to be parsed from a tuple would still benefit higher-level analysis.

If you want to specify the number of employers and you don't already have a settings file, you need to initialize the ABC, call a method that calls _verify_ready() (which either start processing with default number of employers, 50, or rely on some processing to already be completed), call save_settings(), edit the file, re-initialize the object, and call import_settings(). This adds a lot of complexity when trying to change run-time parameters.

**Comment by tjkessler** [comment]
In sample_abc.py, executions must be wrapped in "if __name__ == '__main__':" to avoid multiprocessing errors (double underscores around name and main, markup treats it as bold)

**Comment by tjkessler** [comment]
README has two sections for Installation

**Comment by tjkessler** [comment]
In check_positions(), bees abandon their food sources if their fitness is below the population's average fitness. It's possible that, even though those bees are below the average, through discovering neighboring food sources they can improve. The ABC has the potential to converge on a local minimum instead of a global one. For example if a local minimum is discovered, its fitness could be better than the "edge" or "slope" of a global minimum, causing the bee around the global minimum to abandon it.

From http://www.scholarpedia.org/article/Artificial_bee_colony_algorithm: 
"Employed bees whose solutions cannot be improved through a predetermined number of trials, specified by the user of the ABC algorithm and called “limit” or “abandonment criteria” herein, become scouts and their solutions are abandoned." TL;DR, a bee only abandon a food source if that food source cannot be improved.

**Comment by tjkessler** [comment]
Pep8 styling could use a bit of work:

- 79 character maximum line length (somewhat subjective, but makes reading code on GitHub much easier)
- No "==" for booleans or "None", use "is"
- DocStrings for classes should be under init(), and should outline arguments, return values, etc.
- There's unnecessary white space at the end of lines, between lines
- Spaces are needed between arguments: "np.random.uniform(0,1)" to "np.random.uniform(0, 1)"
- Spaces are needed around operators: "len(self._onlooker.best_employers) -1" to "len(self._onlooker.best_employers) - 1"

I would suggest installing a linter for your IDE of choice, VS Code supports a pep8 linter

**Comment by tjkessler** [comment]
For methods only used by the ABC (e.g. _verify_ready, _gen_probability_values, etc.), use double-underscore naming scheme (especially if these methods are not meant to be called by the user).

Variables should use the double-underscore naming scheme if they should not be accessed by the user. If only the value of the variable should be accessed by the user, an @property method should be used. If the variable should be editable by the user, no underscoring is necessary. See this site for a better explanation of @property methods and setters: https://www.python-course.eu/python3_properties.php

**Comment by tjkessler** [comment]
Having the methods "create_employers" and "calc_new_positions" separate is a good idea - this allows the user to determine a cutoff point for whatever criteria they use, ultimately giving them more control over run-time 👍

**Pull request merged**